src/main/java/io/confluent/connect/hdfs/partitioner/TimeBasedPartitioner.java

```diff
@@ -86,16 +89,53 @@ public void configure(Map<String, Object> config) {
     String hiveIntString = (String) config.get(HdfsSinkConnectorConfig.HIVE_INTEGRATION_CONFIG);
     boolean hiveIntegration = hiveIntString != null && hiveIntString.toLowerCase().equals("true");
 
+    timeFieldName = (String) config.get(HdfsSinkConnectorConfig.PARTITION_TIME_FIELD_NAME_CONFIG);
+
     Locale locale = new Locale(localeString);
     DateTimeZone timeZone = DateTimeZone.forID(timeZoneString);
     init(partitionDurationMs, pathFormat, locale, timeZone, hiveIntegration);
   }
 
   @Override
   public String encodePartition(SinkRecord sinkRecord) {
-    long timestamp = System.currentTimeMillis();
-    DateTime bucket = new DateTime(getPartition(partitionDurationMs, timestamp, formatter.getZone()));
-    return bucket.toString(formatter);
+      long timestamp;
+      if (timeFieldName == null || timeFieldName.equals("")) {
+          timestamp = System.currentTimeMillis();
+      } else if (sinkRecord.value() instanceof Struct) {
+          Struct struct = (Struct) sinkRecord.value();
+          timestamp = getTimestampValue(struct);
+      } else {
+          throw new DataException("sinkRecord parameter is expected to be an instance of Struct");
+      }
+      DateTime bucket = new DateTime(getPartition(partitionDurationMs, timestamp, formatter.getZone()));
+      return bucket.toString(formatter);
+  }
+
+  private long getTimestampValue(Struct record) {
+      final String[] fieldNames = timeFieldName.split("\\.");
+      int i = 0;
+      Struct tmpRecord = record;
+      Object tmpRecordObject = null;
+      while (i < fieldNames.length - 1) {
+          tmpRecordObject = getField(tmpRecord, fieldNames[i]);
+          tmpRecord = (Struct) tmpRecordObject;
+          i++;
+      }
+      tmpRecordObject = getField(tmpRecord, fieldNames[i]);
+      if (!(tmpRecordObject instanceof Long)) {
+          throw new DataException("The value of the time field must be a Long.");
```

Comment:
Would it be valid to check to see if the `tmpRecordObj` is a `Number`, and then call `Number.longValue()`? That way, this logic would work even if the data happens to be encoded as a long, integer, floating point number, or even BigInteger or BigDecimal. It is true that most of the time the value should be a long for semantic reasons, but being a bit more flexible would not seem to hurt.